File: hexa-cli/assets/scaffold/go/composition-root.go
// Package {{name_snake}} is the composition root: the only place allowed to
// name a concrete adapter. Everything else depends on the port.
//
//	  domain  ←  ports  ←  usecases
//	               ↑
//	          adapters (secondary)
//	               ↑
//	   composition-root.go — wires them, once
//
// Check it with `hexa analyze .`.
package {{name_snake}}

import (
	"{{name}}/adapters/secondary"
	"{{name}}/internal/domain"
	"{{name}}/internal/ports"
	"{{name}}/internal/usecases"
)

// Counter builds the application with its real adapters.
//
// The one line below is the whole composition decision. Point it at a
// file-backed store and nothing else moves.
func Counter() ports.CounterStore {
	return secondary.NewInMemoryCounterStore()
}

// IncrementOnce runs the use case against a freshly composed application.
func IncrementOnce() domain.Count {
	return usecases.Increment(Counter())
}

File: hexa-cli/assets/scaffold/ts/src/composition-root.ts
/**
 * {{name}} — a hexagonal skeleton that runs.
 *
 * This is the **composition root**: the only file allowed to name a concrete
 * adapter. Everything else depends on the port.
 *
 *     domain  ←  ports  ←  usecases
 *                  ↑
 *             adapters (secondary)
 *                  ↑
 *      composition-root.ts — wires them, once
 *
 * Check it with `hexa analyze .`.
 */
import { InMemoryCounterStore } from './adapters/secondary/in-memory-counter-store.js';
import type { Count } from './core/domain/count.js';
import type { CounterStore } from './core/ports/counter-store.js';
import { increment } from './core/usecases/increment.js';

/**
 * Build the application with its real adapters.
 *
 * The one line below is the whole composition decision. Point it at a
 * file-backed store and nothing else moves.
 */
export function counter(): CounterStore {
  return new InMemoryCounterStore();
}

/** Run the use case against a freshly composed application. */
export function incrementOnce(): Count {
  return increment(counter());
}
